gtfs/feed_source.py

```diff
@@ -26,10 +38,127 @@ def url(self) -> str:
+                #     LOG.info('GTFS verification succeeded.')
+                #     return True
+                # else:
+                #     LOG.error('GTFS verification failed.')
+                #     return False
+            else:
+                return False
+        else:
+            raise ValueError("URL not set for feed source!")
+
+    def check_header_newer(self, feed_file: str, url: str):
+        """Check if last-modified header indicates a new download is available.
+
+        :param feed_file: Name of downloaded file (relative to :ddir:)
+        :param url: Where GTFS is downloaded from
+        :returns: 1 if newer GTFS available; 0 if info missing; -1 if already have most recent
+        """
+        if not os.path.exists(self.status_file):
+            LOG.debug(f"Status file {self.status_file} not found.")
+            return 0
+
+        with open(self.status_file, "rb") as f:
+            last_status = pickle.load(f)
+        if feed_file in last_status and "posted_date" in last_status[feed_file]:
+            last_fetch = last_status[feed_file]["posted_date"]
+            hdr = requests.head(url)
+            hdr = hdr.headers
+            if hdr.get("last-modified"):
+                last_mod = hdr.get("last-modified")
+                if last_fetch >= last_mod:
+                    LOG.info(f"No new download available for {feed_file}.")
+                    return -1
+                else:
+                    LOG.info(f"New download available for {feed_file}.")
+                    LOG.info(f"Last download from: {last_fetch}.")
+                    LOG.info(f"New download posted: {last_mod}")
+                    return 1
+            else:
+                # should try to find another way to check for new feeds if header not set
+                LOG.debug(f"No last-modified header set for {feed_file} download link.")
+                return 0
+        else:
+            LOG.debug(f"Time check entry for {feed_file} not found.")
+            return 0
+
+    def download_feed(self, feed_file: str, url: str, do_stream: bool = True) -> bool:
+        """Download feed.
+
+        :param feed_file: File name to save download as, relative to :ddir:
+        :param url: Where to download the GTFS from
+        :param do_stream: If True, stream the download
+                :returns: True if download was successful
+        """
+        if self.check_header_newer(feed_file, url) == -1:
+            # Nothing new to fetch; done here
+            return False
+
+        feed_file_path = os.path.join(self.ddir, feed_file + ".zip")
+        LOG.info(f"Getting file {feed_file}...from...{url}")
+        request = requests.get(url, stream=do_stream)
+
+        if request.ok:
+            with open(feed_file_path, "wb") as download_file:
+                if do_stream:
+                    for chunk in request.iter_content(chunk_size=1024):
+                        download_file.write(chunk)
```

Comment:
why stream?

Reply:
l guess since the feeds are pretty big in size, writing them directly is kind of memory inefficient so maybe it's recommended to do so

Reply:
Yes, that's a very good point! Thanks